core/trino-main/src/main/java/io/trino/server/security/oauth2/NimbusOAuth2Client.java

```diff
@@ -0,0 +1,351 @@
+            OIDCTokens tokens = tokenResponse.getOIDCTokens();
+            validateTokens(tokens, nonce);
+            AccessToken accessToken = tokens.getAccessToken();
+            JWTClaimsSet claims = getJWTClaimsSet(accessToken.getValue()).orElseThrow(() -> new ChallengeFailedException("invalid access token"));
+            return new Response(
+                    accessToken.getValue(),
+                    determineExpiration(getExpiration(accessToken), claims.getExpirationTime()),
+                    Optional.ofNullable(tokens.getIDTokenString()));
+        }
+
+        private void validateTokens(OIDCTokens tokens, Optional<String> nonce)
+                throws ChallengeFailedException
+        {
+            try {
+                IDTokenClaimsSet idToken = idTokenValidator.validate(
+                        tokens.getIDToken(),
+                        nonce.map(this::hashNonce)
+                                .map(Nonce::new)
+                                .orElseThrow(() -> new ChallengeFailedException("Missing nonce")));
+                AccessTokenHash accessTokenHash = idToken.getAccessTokenHash();
+                if (accessTokenHash != null) {
+                    AccessTokenValidator.validate(tokens.getAccessToken(), ((JWSHeader) tokens.getIDToken().getHeader()).getAlgorithm(), accessTokenHash);
+                }
+            }
+            catch (BadJOSEException | JOSEException | InvalidHashException e) {
+                throw new ChallengeFailedException("Cannot validate nonce parameter", e);
+            }
+        }
+
+        private String hashNonce(String nonce)
+        {
+            return sha256()
+                    .hashString(nonce, UTF_8)
+                    .toString();
+        }
+    }
+
+    private <T extends AccessTokenResponse> T getTokenResponse(String code, URI callbackUri, NimbusAirliftHttpClient.Parser<T> parser)
+            throws ChallengeFailedException
+    {
+        T tokenResponse = httpClient.execute(new TokenRequest(tokenUrl, clientAuth, new AuthorizationCodeGrant(new AuthorizationCode(code), callbackUri)), parser);
+        if (!tokenResponse.indicatesSuccess()) {
+            throw new ChallengeFailedException("Error while fetching access token: " + tokenResponse.toErrorResponse().toJSONObject());
+        }
+        return tokenResponse;
+    }
+
+    private Optional<JWTClaimsSet> getJWTClaimsSet(String accessToken)
+    {
+        if (userinfoUrl.isPresent()) {
+            return queryUserInfo(accessToken);
+        }
+        return parseAccessToken(accessToken);
+    }
+
+    private Optional<JWTClaimsSet> queryUserInfo(String accessToken)
+    {
+        try {
+            UserInfoResponse response = httpClient.execute(new UserInfoRequest(userinfoUrl.get(), new BearerAccessToken(accessToken)), UserInfoResponse::parse);
+            if (!response.indicatesSuccess()) {
+                return Optional.empty();
```

Comment:
Here we are dropping a lot of errors. I think that when we try to get response from `userInfoUrl`, we should be failing on error. I don't see any failover on `Optional.empty()`, which means that throwing exception here should be fine.

Reply:
I've added error response logging on debug.